### PR TITLE
Respect isolate_session_state in config when ClusterSpec propagation is enabled

### PR DESCRIPTION
**Pull request by shishaochen**
Currently, the `isolate_session_state ` is hard coded to be true when ClusterSpec propagation is enabled.
However, here comes the large-scale distributed training case where workers are running on fragile machines. In this case:
- It is common to see worker N failed and then is restarted on another machine by the infrastructure like Yarn. Then the ClusterSpec held by PS servers are out-of-dated as the IP of worker N has changed.
- It is reasonable that in some time period, the inbound training data increased heavily in size and we want to increase the worker number without restarting the whole job including PS servers.
- It can be helpful to manually replace slow workers detected with new ones on a healthy machine without interrupting the existing training job.

With code change in this pull request, users can decide whether to share state of his Session like Variable values with others through the `isolate_session_state` switch in the session configuration.
In industry, the resource isolation among multiple distributed-training jobs is usually ensured by infrastructure using something like Cgroup or Docker. Hence it is definitely safe and popular to turn off the session isolation for workers inside the same job.

### CODE REVIEW

**Comment by shishaochen** [comment]
I signed it!

**Comment by shishaochen** [comment]
@saeta @jaingaurav Is there any update or feedback? Thanks a lot!

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @jaingaurav, @saeta: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/python/client/session_clusterspec_prop_test.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def testClusterSpecPropagationWorker1Placement(self):
     with ops.Graph().as_default() as g, ops.device('/job:worker/task:0'):
       const = constant_op.constant(17)
     sess = session.Session(server1.target, config=config, graph=g)
-    output = self.evaluate(const)
+    output = sess.run(const)
```
Comment (saeta) [inline tensorflow/python/client/session_clusterspec_prop_test.py]: Quick question: why do you switch from `self.evaluate` to `sess.run`?

File: tensorflow/core/distributed_runtime/session_mgr.cc [hunk 2]
```diff
@@ -76,7 +76,14 @@ Status SessionMgr::CreateSession(const string& session,
 
   std::shared_ptr<WorkerSession> worker_session;
 
-  if (isolate_session_state) {
+  if (isolate_session_state || server_def.cluster().job_size()) {
+    if (server_def.cluster().job_size()) {
+      LOG(INFO) << "ClusterSpec propagation is enabled.";
```
Comment (saeta) [inline tensorflow/core/distributed_runtime/session_mgr.cc]: In general, our users don't like when log messages are printed out that they can't control. Could this be turned to a `VLOG(1)`?

File: tensorflow/python/client/session_clusterspec_prop_test.py [hunk 3]
```diff
@@ -323,8 +325,12 @@ def testClusterSpecPropagationThreeServers(self):
...
     job2.tasks[1] = server3.target[len('grpc://'):]
 
-    config1 = config_pb2.ConfigProto(cluster_def=cluster_def1)
-    config2 = config_pb2.ConfigProto(cluster_def=cluster_def2)
+    config1 = config_pb2.ConfigProto(
+        cluster_def=cluster_def1,
+        isolate_session_state=True)
+    config2 = config_pb2.ConfigProto(
+        cluster_def=cluster_def2,
+        isolate_session_state=True)
```
Comment (saeta) [inline tensorflow/python/client/session_clusterspec_prop_test.py]: If I'm reading this test correctly, it looks like the default will change from previously defaulting to `isolate_session_state=True` to defaulting to `isolate_session_state=False`. Is this true? If so, can we think of a way to avoid this surprising change in behavior?

File: tensorflow/python/client/session_clusterspec_prop_test.py [hunk 4]
```diff
@@ -458,6 +466,86 @@ def testClusterSpecPropagationIsolation(self):
     with self.assertRaises(errors.FailedPreconditionError):
       sess3.run(v)
 
+  def testClusterSpecPropagationNonIsolation(self):
```
Comment (saeta) [inline tensorflow/python/client/session_clusterspec_prop_test.py]: Thanks for adding the test case for the other direction. +1! :-)

**Review by saeta — CHANGES_REQUESTED** [review]

File: tensorflow/python/client/session_clusterspec_prop_test.py (hunk 1 above)
Comment (shishaochen) [reply tensorflow/python/client/session_clusterspec_prop_test.py]: Because, the created Session is not used as a context, so the `ops.get_default_session()` behind the `self.evaluate(...)` will return None and a test Session will be created without the server info.
If `self.evaluate(...)` is preferred, we can use the following instead:
``` python
with session.Session(server1.target, config=config, graph=g) as sess:
  output = self.evaluate(const)
  self.assertEqual(17, output)
```

File: tensorflow/core/distributed_runtime/session_mgr.cc (hunk 2 above)
Comment (shishaochen) [reply tensorflow/core/distributed_runtime/session_mgr.cc]: The code is updated now. Thanks a lot!

File: tensorflow/python/client/session_clusterspec_prop_test.py (hunk 3 above)
Comment (shishaochen) [reply tensorflow/python/client/session_clusterspec_prop_test.py]: @saeta You are right. This incompatibility has embarrassed me a lot :(

AFAIK, in ProtoBuf 3, there is no way to test whether a field is explicitly set or net. Thus, if we set the default value of `isolate_session_state` to be `false`, it will bring another incompatibility.

How about adding a new filed `share_session_state` defaulting to `false` in `tf.ConfigProto` and marking `isolate_session_state` to be deprecated? Then:
- If `share_session_state` is not set, no change happens.
- If `share_session_state` is turned on, a GRPC server will offer renamed devices to its created worker sessions.

Looking forward to your suggestions.

File: tensorflow/python/client/session_clusterspec_prop_test.py (hunk 1 above)
Comment (shishaochen) [reply tensorflow/python/client/session_clusterspec_prop_test.py]: Done with a new [commit](https://github.com/tensorflow/tensorflow/pull/24466/commits/88833ed2e8eaaef44e8c7a5d9a3d9b81c5abb497) which explicitly set the session context.

File: tensorflow/python/client/session_clusterspec_prop_test.py (hunk 3 above)
Comment (shishaochen) [reply tensorflow/python/client/session_clusterspec_prop_test.py]: Done with a new [commit](2bf038739467f5131dd98b91d66747aae6c89972) which adds a new field `share_session_state` in `tf.ConfigProto`.

@saeta @mrry Is this change enough for keeping compatibility? Thanks a lot.

**Comment by shishaochen** [comment]
@saeta @mrry Any update or comment? Thanks a lot.

**Comment by shishaochen** [comment]
@mrry @saeta @jaingaurav @yuefengz Any update or comment? Thanks a lot.

File: tensorflow/core/protobuf/config.proto [hunk 5]
```diff
@@ -388,9 +388,10 @@ message ConfigProto {
   // Optional list of all workers to use in this session.
   ClusterDef cluster_def = 14;
 
-  // If true, any resources such as Variables used in the session will not be
-  // shared with other sessions.
-  bool isolate_session_state = 15;
+  // This option is deprecated. The same effect can be achieved by setting
+  // share_session_state to an inverse value. For compatibility, this field
+  // still works if share_session_state is not turned on.
+  bool isolate_session_state = 15 [deprecated = true];
```
Comment (mrry) [inline tensorflow/core/protobuf/config.proto]: If we deprecate this option, there seems to be no non-deprecated way to isolate session state when ClusterSpec propagation is disabled, which would mean that the new option is not a complete replacement.

Instead of deprecating this field, I would suggest adding a more specific option in the `Experimental` submessage below (e.g. something with a name like "share_session_state_in_clusterspec_propagation"), which will make it easier to merge this change promptly. After that has been merged for a while, the folks working on distributed APIs like `tf.distribute` can make a call about what the long-term supported configuration should be.

Comment (shishaochen) [reply tensorflow/core/protobuf/config.proto]: @mrry Thanks for your suggestions!  
Agree with you that we need a long term API design to control isolation among sessions in distributed runtime.  
Codes are updated and is the pull request qualified now?

**Review by mrry — DISMISSED** [review]
Thanks for making these changes... this looks good to me!

**Comment by shishaochen** [comment]
@saeta I have addressed your requested changes for days.
Could you please help proceed this pull request? Thanks a lot~

**Review by saeta — DISMISSED** [review]
LGTM. Thank you @shishaochen ! (And sorry for the delay on my part.)

File: tensorflow/core/protobuf/config.proto [hunk 6]
```diff
@@ -443,6 +444,12 @@ message ConfigProto {
     // If true, use NCCL for CollectiveOps.  This feature is highly
     // experimental.
     bool collective_nccl = 7;
+    
+    // If true, any resources such as Variables used in the session will be
+    // shared with other sessions even if ClusterSpec Propagation is enabled.
+    // TODO(mrry): This field is a patch for isolate_session_state and an
+    // uniform API with flexibility in distributed runtime is needed.
+    bool share_session_state_in_clusterspec_propagation = 8;
```
Comment (mrry) [inline tensorflow/core/protobuf/config.proto]: In the internal code review, @josh11b commented:

> Please document how this interacts with the isolate_session_state variable more clearly.

Please work with Josh to come up with appropriate wording here!

Comment (shishaochen) [reply tensorflow/core/protobuf/config.proto]: @josh11b @mrry Sorry for late. How about these words as the comment for `
share_session_state_in_clusterspec_propagation`.
```text
// In default, everything works as before when cluster propagation is enabled.
// However, when `share_session_state_in_clusterspec_propagation` is turned on,
// the `isolate_session_state` field will be forced to be true even if you set
// it to false in code. Then, all resources will be shared among worker sessions
// in the same TF Server, just like the case where clusterspec propagation is
// not used in distributed training.
// TODO(mrry): This field is a patch for isolate_session_state and an
// uniform API with flexibility in distributed runtime is needed.
bool share_session_state_in_clusterspec_propagation = 8;
```

If these words satisfy your concerns, I will add them to this PR as a new commit.

**Comment by shishaochen** [comment]
@mrry @josh11b Kindly ping... Appreciate your comments.

File: tensorflow/core/protobuf/config.proto (hunk 6 above)
Comment (josh11b) [reply tensorflow/core/protobuf/config.proto]: I feel like this is too complicated for unstructured prose. How about a bulleted list explaining which combinations of isolate_session_state, share_session_state_in_clusterspec_propagation, and clusterspec propagation lead to isolated vs. shared session state? Like:

* If share_session_state_in_clusterspec_propagation is true and clusterspec propagation is used then session state is shared independent of the value of isolate_session_state
* If ..., then ...
...

Right now you say things like: "when `share_session_state_in_clusterspec_propagation` is turned on, the `isolate_session_state` field will be forced to be true" which just seems backwards and confusing.

Comment (shishaochen) [reply tensorflow/core/protobuf/config.proto]: @josh11b Get the point!
Please review the updated comment. I have pushed the change as a new commit.
Thanks a lot!

File: tensorflow/core/protobuf/config.proto [hunk 7]
```diff
@@ -443,6 +444,27 @@ message ConfigProto {
     // If true, use NCCL for CollectiveOps.  This feature is highly
     // experimental.
     bool collective_nccl = 7;
+    
+    // In the following words, session state means the value of a variable,
```
Comment (shishaochen) [inline tensorflow/core/protobuf/config.proto]: @josh11b Are the following description words clear for uses to understand?
Appreciate your review. Thanks!

**Comment by shishaochen** [comment]
@mrry @saeta 
After a new commit is pushed, your reviews are dismissed in this PR by GitHub.
Could you help review it again? Thanks a lot.

File: tensorflow/core/protobuf/config.proto (hunk 7 above)
Comment (josh11b) [reply tensorflow/core/protobuf/config.proto]: I'd delete the word "words"

Comment (shishaochen) [reply tensorflow/core/protobuf/config.proto]: @josh11b Done :)
The "words" is deleted in the new commit. Thx!

File: tensorflow/core/protobuf/config.proto (hunk 6 above)
Comment (josh11b) [reply tensorflow/core/protobuf/config.proto]: Okay, I'm fine with this description assuming the other reviewers agree that this description actually matches what the code does.

File: tensorflow/core/protobuf/config.proto [hunk 8]
```diff
@@ -443,6 +444,27 @@ message ConfigProto {
     // If true, use NCCL for CollectiveOps.  This feature is highly
     // experimental.
     bool collective_nccl = 7;
+    
+    // In the following, session state means the value of a variable, elements
```
Comment (shishaochen) [inline tensorflow/core/protobuf/config.proto]: @josh11b The "words" is deleted now. Please review it. Thanks!

**Comment by shishaochen** [comment]
@josh11b Kindly ping~

**Comment by shishaochen** [comment]
@mrry @saeta
After a new commit is pushed, your reviews are dismissed in this PR by GitHub.
Could you help review it again? Thanks a lot.

**Comment by shishaochen** [comment]
@josh11b Kindly ping~

**Comment by shishaochen** [comment]
@josh11b Kindly ping~

**Comment by shishaochen** [comment]
@yuefengz Hi, Yuefeng, could you please help invite @josh11b to continue this PR? One week has past TT

**Comment by shishaochen** [comment]
@rthadur Could you please help ping reviewers to continue this PR? Thanks a lot!

**Comment by shishaochen** [comment]
@josh11b Kindly ping~

**Comment by shishaochen** [comment]
@josh11b Kindly ping~

**Comment by shishaochen** [comment]
@josh11b Kindly ping~

**Comment by shishaochen** [comment]
@josh11b Kindly ping~

**Comment by josh11b** [comment]
My last comment was:

"Okay, I'm fine with this description assuming the other reviewers agree that this description actually matches what the code does."

@saeta @mrry  ?

**Review by saeta — APPROVED** [review]

**Comment by shishaochen** [comment]
@josh11b @saeta Thanks a lot for your comments!
Has this PR been accepted after these changes or is there anything I can do to push forward this PR?

**Comment by shishaochen** [comment]
@josh11b @saeta @mrry Kindly ping~

**Comment by saeta** [comment]
Hi @shishaochen Just to let you know, over the past few days, we've been working this PR though the machinery. It's getting closer. Sorry again for the massive delay.
